Stats: Add metric buffering

chrisbliss18 opened the pull request:
Per a request from Systems, this PR adds buffering for StatsD metrics. The buffer flushes when one of two conditions is met:

1. Trying to queue a new metric would cause the buffer to grow larger than the max buffer size.
2. A 5-second interval has elapsed and the buffer is not empty.

The max buffer size is calculated from the MTU size of the network connection that sends UDP packets to the StatsD service. The MTU value is currently hard-coded to a value of `65536` and is overridden to `1500` when running in a development Docker environment. The goal of using the MTU value is to ensure that max packet sizes are used to increase efficiency.

This code has some room for some needed enhancements, primarily the ability to control the host, port, MTU, and flush interval via the config file. For simplicity of testing, and the ability to easily apply this change to production, this enhancement has been left out of this PR. However, if making the config change on production can be easily done, this PR should be updated to allow control of those values via the config file.

### Docker Testing

 - Apply the PR.
 - Run Jetmon.
 - Verify that stats still populate in Graphite (`127.0.0.1:8088`). The following are good metrics to check:
   - For increment, verify that `stats_counts.com.jetpack.jetmon.jetmon.docker.stats.workers.free.count` is updating properly.
   - For timing, verify that `stats.timers.com.jetpack.jetmon.jetmon.docker.db.get_next_batch.count` is updating properly.

### Production Server Testing

 - Follow the testing instructions provided by bisko at pMz3w-h0a-p2#comment-106587 to run a local test with `node test.js`. Modify the example code to have the timeout that is long enough to ensure that the 5-second delayed flush can execute.
 - Verify that the test metric shows up in the StatsD data.

darssen commented:
Thanks for working on this!

I really like the idea of the `init` method to have more control of the configuration and the flow itself. I too agree that we could enhance the configuration further, but it is a good idea to have it in another PR.

Now when testing in Docker I found some logs that indicated something is not working as expected. 

Getting a bunch of this in the statsd container

> docker-statsd-1     | 3 Mar 10:26:57 - DEBUG: Bad line: 1 in msg "6|m"
docker-statsd-1     | 3 Mar 10:26:57 - DEBUG: Bad line: 1 in msg "|@1"
docker-statsd-1     | 3 Mar 10:26:57 - DEBUG: Bad line: 1 in msg "6|"
docker-statsd-1     | 3 Mar 10:26:57 - DEBUG: Bad line: 1 in msg "com.jetpack.jetmon.jetmon.docker.worker.check.up.rt"
docker-statsd-1     | 3 Mar 10:26:57 - DEBUG: Bad line: 1 in msg "6|ms|@1"
docker-statsd-1     | 3 Mar 10:26:57 - DEBUG: Bad line: 1 in msg "@1"
docker-statsd-1     | 3 Mar 10:26:57 - DEBUG: Bad line: 14 in msg "com.jetpack.jetmon.jetmon.docker.worker.check.up.rtt:14"

First I confirmed this is not happening in `master` and then, I tried to investigate further and saw that in `Graphite` I was getting some weird strings
![image](https://user-images.githubusercontent.com/13239096/222703363-f8733296-d679-45fd-b0d2-36a1fb5b5f02.png)

I think there is something happening with the slicing in line 91 of statsd.js
```javascript
const buffer = this.buffer.slice( 0, this.bufferLength )
```
Not sure what is the problem, but after a very naive `toString()` conversion, DEBUG errors are no longer showing, and reports still show in Graphite. This might give you a hint.

chrisbliss18 commented:
@darssen that's interesting that you had that issue. I experienced that at one point in testing. After making some changes (not targeting that issue), it went away. I'll work on recreating the issue, and then I'll test out using a `toString()` conversion.

chrisbliss18 commented:
The implementation of the enhancement has been changed due to feedback from numerous parties. The following describes the key changes:

- Each worker no longer sends StatsD metrics. This improves efficiency as each worker no longer has its own `statsdClient` buffer.
- Each worker keeps track of the http code stats in aggregate and the rtt stats as just avg, max, and min values. These aggregate stats are sent to the parent process as part of the pre-existing per-second stats message. This improves efficiency as now all `statsdClient` communication is centralized to the parent process.
- The main `jetmon.js` code's `updateStats()` function now sends the aggregated worker check stats. Since this function executes each `STATS_UPDATE_INTERVAL_MS`, this means that the frequency of sending these aggregate stats is now controllable by that config.
- Removed use of a `Buffer` object from the `statsdClient` code. I added it to try to squeeze out extra memory optimizations since each worker also ran its own buffer, but it clearly created odd issues as noted by @darssen above. So, now it's just a normal string.

darssen commented:
I am able to run the `statsd` container now without any issues.

darssen commented on lib/jetmon.js:
```diff
@@ -488,6 +511,18 @@ function updateStats() {
 
+		for ( let site_status in checkStats ) {
+			for ( let http_code in checkStats[site_status]['http_code'] ) {
+				statsdClient.increment( `worker.check.${site_status}.code.${http_code}.count`, checkStats[site_status]['http_code'][http_code] );
```
Since all these stats are not per worker anymore, does it make sense to name it this way?

Maybe something like `'stats.check.${site_status}.code.${http_code}.count'`

Same applies for lines below

bisko replied:
I think it's fine to keep them as `worker`, as they come from the workers.
